### ISSUE TITLE
Problem parsing body: No signatures found matching the expected signature for payload

### ISSUE THREAD

**Issue by mcrider** [stripe/stripe-node#356]
I'm using AWS API Gateway/Lambda which passes the request body as an actual JSON object to my Lambda function, where I am calling `stripe.webhooks.constructEvent` on it.  It seems this function expects the payload to be stringified JSON, so I run `JSON.stringify` on the request body, but this fails to sign (I'm guessing because JSON.stringify subtly changes what the JSON string would look like coming from the server).  Is it possible for constructEvent to detect if the payload is a JSON object and skip parsing it?  I mean, in a way that maintains security?

**Comment by itai-codefresh** [comment]
actually from the documentation you need to pass the buffered body content.
but i am having the same trouble, using exactly the documented solution.
it doesn't work. i also tried to look at the inner code that performs the validation and i don't see what is the problem. maybe a different hash function that stripe is using than node

**Comment by jlomas-stripe** [comment]
You need to pass in the _exact_ raw text body that was received from Stripe, rather than a re-`stringify`'d version of it.

@mcrider I'm not super familiar with Lambda, but I think you want to do something like this: `"body": "$input.body"`: http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#input-variable-reference

@itai-codefresh If you're using Express, you can just use some [vanilla middleware](https://github.com/stripe/stripe-node/blob/master/examples/webhook-signing/express.js) to get the raw body for the request.

**Comment by itai-codefresh** [comment]
you are right @jlomas-stripe it was my bad

**Comment by mcrider** [comment]
Thanks guys -- Its a little tricky getting the raw body to your Lambda function but I understand why its necessary!  I'll go ahead and close this.

**Comment by jlomas-stripe** [comment]
@itai-codefresh No worries - it's a bit tricky, and my initial examples caused some fun `body-parser` collisions so it's fun, fun stuff. :)

@mcrider Awesome.  It'd be awesome if you can drop by and leave your solution once you find one! :)

**Comment by mcrider** [comment]
Figured it out!

- In the AWS admin, go to your API gateway endpoint for your webhook, then go to the integration request.
- Open the Body Mapping Templates section then select application/json -- Scroll down and you'll see your body mapping template
- Add this line: `"rawbody": "$util.escapeJavaScript($input.body)",`
- I'm using Serverless, and needed to redeploy my whole stack.  You'll have to do similar with whatever tool you're using to get API gateway to send the rawbody attribute (this didn't seem necessary for me for other mappings like IP address, so not sure whats going on)
- In your Lambda function, you should be able to use event.rawbody as is, without unescaping.  Here's my code (node.js)
```
var endpointSecret = "whsec_XXXXXXXX";
var headers = JSON.parse(event.headers);
var stripeEvent = stripe.webhooks.constructEvent(event.rawbody, headers["Stripe-Signature"], endpointSecret);
```

I was just going to filter requests by IP (checking against the list of Stripe IPs) but this is probably a more future-proof technique.

**Comment by jlomas-stripe** [comment]
@mcrider Awesome. :)  Just so you're aware, it sounds like `$util.escapeJavaScript()` might [misbehave for certain characters](https://forums.aws.amazon.com/thread.jspa?threadID=228917) so you may just want to consider doing `"rawbody": "$input.body",` instead; if `escapeJavaScript()` does happen to actually escape something, the signature of the (escaped) body won't match, so `constructEvent()` would fail.

Since it's unlikely that the event will both have a valid signature AND contain some kind of nefarious JavaScript, I think it's probably safe to just use the raw input - but that's of course up to you. :)

**Comment by mcrider** [comment]
I definitely tried, but I got fatal errors when I called the endpoint (I guess because the unescaped strings would make the JSON sent to the lambda function invalid). Maybe I should just replace quote characters with escaped quotes instead of doing a full escape as suggested in that thread. But futzing with this is a PIA and I'm going to put this on the backburner for now :)

**Comment by alexcasalboni** [comment]
@mcrider I thought the LAMBDA_PROXY integration would always include the raw body. Isn't `event.body` enough? I'm experiencing the same issue (I'm also using the Serverless Framework). Did you actually change the LAMBDA_PROXY integration?

**Comment by mcrider** [comment]
I think this is really a YMMV situation, but I am not using Lambda Proxy integration.  I realized that the Serverless framework, at least how I had it configured, passes along a body mapping template (as it should, you don't want to have to set this sort of thing in the AWS console where its not easily repeatable), so I was having a weird interaction of how Serverless configured the API gateway and what I was setting in the AWS console.  If you're using another technology to set up API Gateway/Lambda it might have its own behavior you need to be aware of.

**Comment by alexcasalboni** [comment]
What version of the Serverless Framework are you using? The latest version
should use proxy integration by default.

**Comment by mcrider** [comment]
1.0.3, so fairly old.

**Comment by alexcasalboni** [comment]
Yeah, that version is [9 months old](https://github.com/serverless/serverless/tree/v1.0.3) :)

I'd recommend migrating to the latest stable version (1.16.1), although it looks like I am experiencing a similar problem with v1.15.0!

**Comment by sunlee-newyork** [comment]
Hi, what solution can I use for node (Meteor) and DigitalOcean?

I'm getting the request body as a Buffer:
```
req.body:  <Buffer 7b 0a 20 20 22 63 72 65 61 74 65 64 22 3a 20 31 33 32 36 38 35 33 34 37 38 2c 0a 20 20 22 6c 69 76 65 6d 6f 64 65 22 3a 20 66 61 6c 73 65 2c 0a 20 20 ... >
```
I am using a middleware called Picker that uses body-parser, so I am initializing as suggested in the Stripe doc:
```
Picker.middleware(bodyParser.raw({type: "*/*"}));
```
But I am getting the same `Error: No signatures found matching the expected signature for payload` message.

**Comment by jlomas-stripe** [comment]
@sunlee-newyork Can you share the code you're using in your webhook handler where you're trying to process the `request.body`?

**Comment by sunlee-newyork** [comment]
Hi @jlomas-stripe, I fixed the issue a few days ago by resorting to my own custom HMAC SHA comparison as described in the stripe docs. Unfortunately I don't recollect the issue...

**Comment by jlomas-stripe** [comment]
Ahhh, ok.  Glad to hear it!

**Comment by jaynsw** [comment]
The example is wrong.The endpoint security key should be a buffer type.

let event = stripe.webhooks.constructEvent(req.body, sig, Buffer.from(endpointSecret));

**Comment by jlomas-stripe** [comment]
@jaynsw Actually, it can be [either a string or a Buffer](https://github.com/stripe/stripe-node/blob/55ebb450b2944396ad0d202891c5ae11c0d89233/lib/Webhooks.js#L31) - it doesn't matter which you use.

**Comment by catamphetamine** [comment]
@mcrider Don't forget the single quote escaping, otherwise it will throw `400 Bad Request`.
https://forums.aws.amazon.com/thread.jspa?messageID=900916&#900916
```js
$util.escapeJavaScript(data).replaceAll("\\'","'")
```

> This function will turn any regular single quotes (') into escaped ones (\'). However, the escaped single quotes are not valid in JSON. Thus, when the output from this function is used in a JSON property, you must turn any escaped single quotes (\') back to regular single quotes ('). This is shown in the following example:

**Comment by idhard** [comment]
hey i now is a closed and old issue , but i'm facing the same problem with serverless , how did you managed to pass raw body with proxy integration ?  i can't pass signature validation :S , thanks

**Comment by jlomas-stripe** [comment]
@idhard If you [reach out to Support](https://support.stripe.com/contact/) - along with more details on your implementation - they can help you with that.

**Comment by comfytoday** [comment]
I'm using AWS Amplify with AWS Serverless Express via API Gateway to Lambda using Lambda Integration.

Hope someone else finds this userful.

I managed to get webhooks working like this:



```
app.use(bodyParser.raw({ type: '*/*' }))

app.post(
  '/webhook',
  (request, response) => {
try {
      event = stripe.webhooks.constructEvent(request.body, request.headers['stripe-signature'], 'whsec...')
    } catch (err) {
      // invalid signature
      console.log('err constructing event: ', err)      
      return response.status(400).send(`Webhook Error: ${err.message}`)
    }
})

```

**Comment by abtx** [comment]
Same issue here. I'm using Lambdas via Netlify, and I don't seem to have the app in AWS Admin where I could set up Body Mapping Templates. Is there any other way I can get the raw response from the event?

**Comment by remi-stripe** [comment]
@abtx I'd recommend talking to Netlify's support team directly asking for help on this and possibly shipping a change to support this

**Comment by akanksha1909** [comment]
I am using npm stripe version 3.5.0(old version probably 4 years ago) and while verifying the webhook signature, i am getting error **TypeError: Cannot read property 'constructEvent' of undefined**. I am using rawBody. i am not sure is the stripe npm version is problem or something else? 
```
function getRawBody(req, res, next) {
  req.rawBody = '';

  req.on('data', function (chunk) {
    req.rawBody += chunk;
  });
  req.on('end', function () {
    next();
  });
}

```
```
app.post('/shook',getRawBody, (request, response) => {
  const sig = request.headers['stripe-signature'];
  let event;
  try {
    event = stripe.webhhoks.constructEvent(request.rawBody, sig, endpointSecret);
  } catch (err) {
    return response.status(400).send(`Webhook Error: ${err.message}`);
  }

  // Handle the checkout.session.completed event
  if (event.type === 'checkout.session.completed') {
    const session = event.data.object;
  
    //Complete function here ...
  }

  // Return a response to acknowledge receipt of the event
  response.json({received: true});
});

```

**Comment by remi-stripe** [comment]
@mcrider this method was added in 4.19.0 as documented in the [changelog](https://github.com/stripe/stripe-node/blob/master/CHANGELOG.md#4190---2017-05-11) so it won't work with your version of stripe-node. You'll need to upgrade to a newer version for the code to work.

**Comment by defionscode** [comment]
I have a thorough fix that I figured out with @ambergkim

First configure the mapping template as you all had already discussed

**mapping template in api gateway**

```js
#set($allParams = $input.params())
{
"raw": "$input.body",
"body-json" : $input.json('$'),
"params" : {
#foreach($type in $allParams.keySet())
    #set($params = $allParams.get($type))
"$type" : {
    #foreach($paramName in $params.keySet())
    "$paramName" : "$util.escapeJavaScript($params.get($paramName))"
        #if($foreach.hasNext),#end
    #end
}
    #if($foreach.hasNext),#end
#end
},
"stage-variables" : {
#foreach($key in $stageVariables.keySet())
"$key" : "$util.escapeJavaScript($stageVariables.get($key))"
    #if($foreach.hasNext),#end
#end
},
"context" : {
    "account-id" : "$context.identity.accountId",
    "api-id" : "$context.apiId",
    "api-key" : "$context.identity.apiKey",
    "authorizer-principal-id" : "$context.authorizer.principalId",
    "caller" : "$context.identity.caller",
    "cognito-authentication-provider" : "$context.identity.cognitoAuthenticationProvider",
    "cognito-authentication-type" : "$context.identity.cognitoAuthenticationType",
    "cognito-identity-id" : "$context.identity.cognitoIdentityId",
    "cognito-identity-pool-id" : "$context.identity.cognitoIdentityPoolId",
    "http-method" : "$context.httpMethod",
    "stage" : "$context.stage",
    "source-ip" : "$context.identity.sourceIp",
    "user" : "$context.identity.user",
    "user-agent" : "$context.identity.userAgent",
    "user-arn" : "$context.identity.userArn",
    "request-id" : "$context.requestId",
    "resource-id" : "$context.resourceId",
    "resource-path" : "$context.resourcePath"
    }
}
```

Then what happens is that `event.raw` is base64 encoded, so when you try to send that to stripe it will fail.

<img width="876" alt="image" src="https://user-images.githubusercontent.com/4664162/74059538-39c04980-49b6-11ea-8cc4-343f1b7c3654.png">

Then in your lambda

```js
// function handler
exports.handler = async (event, context) => {
  // this decodes the b64 into a buffer
  const rawBodyAsBuffer = (new Buffer(event.raw, 'base64'))
  try {
    // just borrowed from stripe docs, obv there are undefined vars in this example
    event = stripe.webhooks.constructEvent(rawBodyAsBuffer, sig, endpointSecret);
  }
  catch (err) {
    response.status(400).send(`Webhook Error: ${err.message}`);
  }
  return {
    statusCode: 200
  }
}
```

And you should be all set.

**Comment by r0fls** [comment]
I don't see mapping templates in my aws lambda gateway configuration 😢 Also none of the express solutions seem to be working

**Comment by ambergkim** [comment]
@r0fls if you are willing to try other tools, the architect framework passes the body as base 64 encoded depending on how you access the body through the framework. You can then create a new buffer from there.

**Comment by chaunceyau** [comment]
I had this same issue. If using lambda proxy (the default/recommended) integration - you need to simply pass the `event.body` object to the function. It is **already stringified**.. The rest of the object is not, but the `body` is.

```
  module.exports.hello = (event) => {
    stripe.webhooks.constructEvent(event.body, sig, endpointSecret)
  }
```

**Comment by michelalbers** [comment]
@chaunceyau This did not work for me :(

An easy solution is to get the id from the payload and then retrieve the event from stripe:

```TypeScript
const payload = JSON.parse(event.body);
const stripeEvent = await stripe.events.retrieve(payload.id);
```

**Comment by lmanerich** [comment]
> @chaunceyau This did not work for me :(
> 
> An easy solution is to get the id from the payload and then retrieve the event from stripe:
> 
> ```ts
> const payload = JSON.parse(event.body);
> const stripeEvent = await stripe.events.retrieve(payload.id);
> ```
> 
> That way you can trust the event as well.

Doing this way we may end up granting the same "purchase" multiple time.

**Comment by ahallora** [comment]
> I had this same issue. If using lambda proxy (the default/recommended) integration - you need to simply pass the `event.body` object to the function - **WITHOUT ANY PARSING/BUFFER STEPS**. It is **already stringified**.. The rest of the object is not, but the `body` is.
> 
> ```js
>   module.exports.hello = (event) => {
>     stripe.webhooks.constructEvent(event.body, sig, endpointSecret)
>   }
> ```

Just to elaborate a bit on this solution, I was getting this error `Unable to extract timestamp and signatures from header` and had to strip the `stripe-signature` header of `[` and `]` like this:
```javascript
module.exports.hello = (event) => {
    const sig = (event.headers["Stripe-Signature"] || "").replace(/\[|]/gi, ""); 
    const stripeEvent = stripe.webhooks.constructEvent(event.body, sig, endpointSecret);
}
```

**Comment by vdanyliv** [comment]
for those who use serverless framework:

your function should look like this.
```
functions:
  webhook:
    handler: Webhook/index.handler
    timeout: 30
    events:
      - http:
          path: /api/billing/webhook
          method: POST
          integration: lambda
          cors: true
          request:
            template:
              application/json:
                '{"rawBody": "$util.base64Encode($input.body)",
                  "headers": {
                    #foreach($param in $input.params().header.keySet())
                    "$param": "$util.escapeJavaScript($input.params().header.get($param))"
                    #if($foreach.hasNext),#end
                    #end
                }}'
```

After your create your function, you need to parse base64 and convert to utf8 string;

`const raw = Buffer.from(rawBody, 'base64').toString('utf8');`

Now you can use it for verification

```
const stripeEvent = stripe.webhooks.constructEvent(
  raw,
  headers['Stripe-Signature'],
  process.env.whk,
);
```

**Comment by fr0gs** [comment]
Hello @vdanyliv I am right now having the same issue but I can't seem to make it work.

**serverless.yml**
```yml
functions:
  main:
    handler: handler.main
    events:
      - http:
          path: verify
          method: post
          integration: lambda
          cors: true
          request:
            template:
              application/json:
                '{"rawBody": "$util.base64Encode($input.body)",
                  "headers": {
                    #foreach($param in $input.params().header.keySet())
                    "$param": "$util.escapeJavaScript($input.params().header.get($param))"
                    #if($foreach.hasNext),#end
                    #end
                }}'
```

**handler.js**
```js
module.exports.main = async ({body, headers}) => {
  
  const verified = utils.verifyRequestSignature(body, headers);

  console.log(verified);

  return {
    statusCode: 200,
  };
```

**utils.js**
```js
function verifyRequestSignature(body, headers) {

    const rawPayload = Buffer.from(body, 'base64').toString('utf8');

    console.log(rawPayload);
    console.log(headers['Stripe-Signature']);

    let stripeEvent;
    
    try {
      const stripeEvent = stripe.webhooks.constructEvent(
        rawPayload,
        headers['Stripe-Signature'],
        process.env.STRIPE_WEBHOOK_SECRET
      );
    }
    catch (err) {
      console.log(`Webhook Error: ${err.message}`);
      return false;
    }
```


And *verified* is still evaluating to false, constantly throwing also the error:
```
3839d608-4899-4199-a11s-20785858ef4c	INFO	Webhook Error: No signatures found matching the expected signature for payload. Are you passing the raw request body you received from Stripe? https://github.com/stripe/stripe-node#webhook-signing
```

Am I missing something else? thank you

**Comment by thomasstep** [comment]
I'm using an API Gateway/Lambda integration and my call is accepting the `event.body` without any buffering, remapping, or parsing.

```javascript
exports.handler = async function (event, context, callback) {
  const webhookSecret = process.env.STRIPE_WEBHOOK_SECRET;
  const sig = event?.headers['Stripe-Signature'];
  const stripeEvent = stripe.webhooks.constructEvent(event.body, sig, webhookSecret);
  ...
```

**Comment by fathur** [comment]
I have this issue too, after long investigation it caused by my middleware.

Unfortunately I use middleware [middy jsonBodyParser](https://www.npmjs.com/package/@middy/http-json-body-parser), and it caused `event.body` which is formatted string json, becomes json object.

If you want to debugging the error, I suggest you must see the `event.body` content. The content has to be EXACTLY SAME as stripe was sent. And the format is "**string** of formatted json". 

What I mean is formatted string below
```
'{
  "foo": {
    "bar": 12
  }
}'
```
is different with this
```
'{"foo": {"bar": 12}}'
```

**Comment by atul-chaudhary** [comment]
> Figured it out!
> 
> * In the AWS admin, go to your API gateway endpoint for your webhook, then go to the integration request.
> * Open the Body Mapping Templates section then select application/json -- Scroll down and you'll see your body mapping template
> * Add this line: `"rawbody": "$util.escapeJavaScript($input.body)",`
> * I'm using Serverless, and needed to redeploy my whole stack.  You'll have to do similar with whatever tool you're using to get API gateway to send the rawbody attribute (this didn't seem necessary for me for other mappings like IP address, so not sure whats going on)
> * In your Lambda function, you should be able to use event.rawbody as is, without unescaping.  Here's my code (node.js)
> 
> ```
> var endpointSecret = "whsec_XXXXXXXX";
> var headers = JSON.parse(event.headers);
> var stripeEvent = stripe.webhooks.constructEvent(event.rawbody, headers["Stripe-Signature"], endpointSecret);
> ```
> 
> I was just going to filter requests by IP (checking against the list of Stripe IPs) but this is probably a more future-proof technique.

Followed the same approach for retrieving the raw body by adding the line to a mapping template in the API gateway.
here is my mapping template
```
##  See http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html
##  This template will pass through all parameters including path, querystring, header, stage variables, and context through to the integration endpoint via the body/payload
#set($allParams = $input.params())
{
"rawbody": "$util.escapeJavaScript($input.body).replaceAll("\\'","'")",    //[tried using without  .replaceAll("\\'","'") result is same]/
"body-json" : $input.json('$'),
"params" : {
#foreach($type in $allParams.keySet())
    #set($params = $allParams.get($type))
"$type" : {
    #foreach($paramName in $params.keySet())
    "$paramName" : "$util.escapeJavaScript($params.get($paramName))"
        #if($foreach.hasNext),#end
    #end
}
    #if($foreach.hasNext),#end
#end
},
"stage-variables" : {
#foreach($key in $stageVariables.keySet())
"$key" : "$util.escapeJavaScript($stageVariables.get($key))"
    #if($foreach.hasNext),#end
#end
},
"context" : {
    "account-id" : "$context.identity.accountId",
    "api-id" : "$context.apiId",
    "api-key" : "$context.identity.apiKey",
    "authorizer-principal-id" : "$context.authorizer.principalId",
    "caller" : "$context.identity.caller",
    "cognito-authentication-provider" : "$context.identity.cognitoAuthenticationProvider",
    "cognito-authentication-type" : "$context.identity.cognitoAuthenticationType",
    "cognito-identity-id" : "$context.identity.cognitoIdentityId",
    "cognito-identity-pool-id" : "$context.identity.cognitoIdentityPoolId",
    "http-method" : "$context.httpMethod",
    "stage" : "$context.stage",
    "source-ip" : "$context.identity.sourceIp",
    "user" : "$context.identity.user",
    "user-agent" : "$context.identity.userAgent",
    "user-arn" : "$context.identity.userArn",
    "request-id" : "$context.requestId",
    "resource-id" : "$context.resourceId",
    "resource-path" : "$context.resourcePath"
    }
}
```
And also parsing headers like ```var headers = JSON.parse(event.headers);   //line 26```

but when i send a test event from stripe webhook implementation it through 
```
{
  "errorType": "SyntaxError",
  "errorMessage": "Unexpected token u in JSON at position 0",
  "trace": [
    "SyntaxError: Unexpected token u in JSON at position 0",
    "    at JSON.parse (<anonymous>)",
    "    at Runtime.exports.handler (/var/task/index.js:26:24)",
    "    at Runtime.handleOnce (/var/runtime/Runtime.js:66:25)"
  ]
}
```
Stuck with this problem for a long time.

**Comment by IcebergRage** [comment]
> > @chaunceyau This did not work for me :(
> > An easy solution is to get the id from the payload and then retrieve the event from stripe:
> > ```ts
> > const payload = JSON.parse(event.body);
> > const stripeEvent = await stripe.events.retrieve(payload.id);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That way you can trust the event as well.
> 
> Doing this way you may end up granting the same "purchase" multiple time.

I'm trying to make it works with `stripe.webhooks.constructEvent` but it's always failing. With `stripe.event.retrieve`, it works well. 

Someone can explain why it's not recommended ?

**Comment by redouane-dev** [comment]
**TL;DR** make sure you are using the signing secret from the Stripe dashboard (either test or live mode), and not the secret obtained from the CLI with `stripe listen --forward-to xxx`. They are different.

I had the same issue for 2 days and tried the solutions in this thread, but it had nothing to do with the API GW body parsing. I was using the signing secret returned by the Stripe CLI (when doing `stripe listen --forward-to`), and testing locally (with the serverless-offline plugin) worked perfectly, so I went live on my staging Lambda and there the signing failed - because the secret was not the right one ofc.

Using the Serverless framework, and API Gateway with a LAMBDA integration - so no need to set a request template.

```yaml
// serverless.yaml

...
functions:
  webhook:
    name: projectname-stage-webhook
    description: xxx
    handler: path/to/handler
    role: webhookFunctionRole
    events:
      - http:
          path: webhook
          method: post
          cors: false
...
```

```ts
// handler.ts

const rawBody = event.body;
const requestSignature = event.headers['Stripe-Signature']; // mind the casing
const webhookSecret = 'whsec_xxx', // make sure this is coming from the Stripe dashboard when deploying to your Lambda

try {
  return stripe.webhooks.constructEvent(
    rawBody,
    requestSignature,
    webhookSecret,
  );
} catch (error) {
  log.error('Webhook signature verification failed.', {
    error,
    request: {
      body: rawBody,
      signature: requestSignature,
    },
  });
  throw new StatusCodeError(400, 'Signature verification failed');
}
```

**Comment by ioveracker** [comment]
Apologies for resurrecting this thread, but it took me a lot of trial and error with the solutions explained here to find what works.

## Serverless Framework folks

[As mentioned above](https://github.com/stripe/stripe-node/issues/356#issuecomment-881277703), using `middy` and `middyJsonBodyParser` in your Serverless Framework lambda can be the root cause of a signature mismatch. I can confirm that after removing this middleware, I can pass `event.body` as-is into `stripe.webhooks.constructEvent` without any request template / rawbody steps. I simply removed the `middy` wrapper around my handler function and the signature is now being verified.

So, if you have this in your handler, remove it and just export your handler function:

```
export const main = middy(webhook).use(middyJsonBodyParser());
```

`serverless` v2.59.0
`stripe` v9.4.0

**Comment by millin-stripe** [comment]
You can also try this. The return to the API end point is coming in is Base64 encoded. So all you need to do is decode it. 

exports.handler = async (event, context) => {
    //Get the webhook body and headers. Body is base64encoded
    let buff = Buffer.from(event.body, "base64");
    let eventBodyString = buff.toString('UTF-8');
    var eventBody = JSON.parse(eventBodyString);
    let endpointSecret =  "yoursec"
    let sig = eventHeaders['Stripe-Signature'];
    let eventHeaders = event['headers'];

Once that is done you can use the SDK as normal. 
stripeEvent = stripe.webhooks.constructEvent(eventBodyString, sig, endpointSecret);

**Comment by codercor** [comment]
> ```js
>  const rawBodyAsBuffer = (new Buffer(event.raw, 'base64'))
> ```

Your API Gateway is "HTTP API"  you should create "rest API" instead of it.

**Comment by jkups** [comment]
> I'm using AWS Amplify with AWS Serverless Express via API Gateway to Lambda using Lambda Integration.
> 
> Hope someone else finds this userful.
> 
> I managed to get webhooks working like this:
> 
> ```
> app.use(bodyParser.raw({ type: '*/*' }))
> 
> app.post(
>   '/webhook',
>   (request, response) => {
> try {
>       event = stripe.webhooks.constructEvent(request.body, request.headers['stripe-signature'], 'whsec...')
>     } catch (err) {
>       // invalid signature
>       console.log('err constructing event: ', err)      
>       return response.status(400).send(`Webhook Error: ${err.message}`)
>     }
> })
> ```

While the above works, note that `app.use(bodyParser.raw(...))` will parse all request body as raw regardless of the endpoint. It is generally recommended to assign parsers per endpoint rather than app wide. So if you've got multiple endpoints with different parsing requirements then you may want to do this instead.

```
const jsonParser = bodyParser.json()
const rawParser = bodyParser.raw({ type: '*/*' })

app.post(jsonEndpoint, jsonParser, (req, res) => {})
app.post(rawEndpoint, rawParser, (req, res) => {})
```
I think you can also use `express.json()` and `express.raw()` as they are based on bodyParser. You will need to confirm that though.

**Comment by tmokmss** [comment]
For people using Next.js with AWS Lambda, I confirmed the below code works.

```ts
import getRawBody from 'raw-body';
import { NextApiRequest, NextApiResponse } from 'next';

export const config = {
  api: {
    bodyParser: false,
  },
};

const webhookHandler = async (req: NextApiRequest, res: NextApiResponse) => {
  const buf = await getRawBody(req);
  const sig = req.headers['stripe-signature'] as string;

  const event = stripe.webhooks.constructEvent(buf, sig, 'whsec_xxx');
  // ...
}
```

**Comment by notturingtested** [comment]
I am still having this issue, I confirmed my keys are correct. Here is my code and the CDK to deploty my lambda:

`exports.handler = async function (event) {
  const stripe = require("stripe")(process.env.STRIPE_SECRET, {
    apiVersion: "2020-08-27",
  });
  const webhookSecret = process.env.STRIPE_WEBHOOK_SECRET;
  const requestSignature = event.headers['Stripe-Signature']; // mind the casing
  let stripeEvent
  try {
    stripeEvent = stripe.webhooks.constructEvent(
      event.body,
      requestSignature,
      webhookSecret,
    );
    } catch (err) {
      // On error, log and return the error message.
      console.log(`❌ Error message: ${err.message}`);
      return { statusCode: 400, body: `Webhook Error: ${err.message}` };
    }
}`


`createNewLambda(scope: Construct, props: YPMLambdaProps): Function{
    return new Function(
      scope,
      props.name,
      {
        runtime: Runtime.NODEJS_18_X,
        handler: 'index.handler',
        code: Code.fromAsset(
          path.join(__dirname, props.relativeCodePath),
        ),
        functionName: props.name,
        role: props.role ? props.role : this.defaultLambdaRole,
        timeout: Duration.seconds(10),
        architecture: Architecture.ARM_64,
      },
    );

  }
  
    createNewLambdaExecutionRole(scope: Construct, props: YPMLambdaExecutionRoleProps): Role {
    const lambdaExecutionRole = new Role(scope, props.name, {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      description: props.description,
    });
    lambdaExecutionRole.addManagedPolicy(
      ManagedPolicy.fromAwsManagedPolicyName(
        'service-role/AWSLambdaBasicExecutionRole',
      ),
    );
    return lambdaExecutionRole
  }`

**Comment by lmanerich** [comment]
> stripeEvent = stripe.webhooks.constructEvent(
>       event.body,
>       requestSignature,
>       webhookSecret,
>     );

Hey @notturingtested you need to read the rawBody attr from event.

**Comment by ice-lenor** [comment]
@redouane-dev Thank you, you saved me probably hours of googling. It never would've occurred to me that I was using the wrong secret. It is too easy to copy-paste settings from local dev to lambda... Thanks!

**Comment by Che-headway** [comment]
> [@mcrider](https://github.com/mcrider) I thought the LAMBDA_PROXY integration would always include the raw body. Isn't `event.body` enough? I'm experiencing the same issue (I'm also using the Serverless Framework). Did you actually change the LAMBDA_PROXY integration?





https://preprocessor.space/questions/using-php-to-parse-a-string-in-json-format

**Comment by mbeato** [comment]
since this thread is still the top google hit for "stripe webhook signature mismatch" — quick brain dump of the three failure modes i kept hitting:

1. **body parsed before signing.** express needs `express.raw({type:'application/json'})` mounted *before* `express.json()` on the webhook route. lambda needs `event.body` not the parsed object. `JSON.stringify` reorders keys / escapes differently from what stripe signed, so the rebuilt body never hashes back.

2. **wrong secret type.** `whsec_*` is the webhook secret (dashboard → developers → webhooks → reveal). `sk_*` is your api secret, `pk_*` is the publishable key. they're not interchangeable and the error doesn't say which one you passed.

3. **timestamp skew after a `stripe-cli` replay** — default tolerance is 5min, replays can blow past it.

if it'd save anyone time i made [stripesig](https://stripesig.apimesh.xyz?utm_source=github&utm_medium=issue-comment&utm_campaign=stripe-node-356) that runs through these automatically — paste payload + secret + signature header and it explains which case it is. open source, no signup, also handles github/slack/shopify hmac.
